py/dml/codegen.py

```diff
@@ -943,6 +944,101 @@ def add(self, method):
 
 simple_events = SimpleEvents()
 
+def check_designated_initializers(site, etype, init_asts):
+    shallow_real_etype = safe_realtype_shallow(etype)
+    duplicates = set()
+    bad_fields = set()
+    remaining = set(shallow_real_etype.members)
+    for (field, init) in init_asts:
+        if field not in shallow_real_etype.members:
+            bad_fields.add(field)
+        elif field not in remaining:
+            duplicates.add(field)
+        else:
+            remaining.remove(field)
+    if (duplicates or bad_fields or remaining):
+        raise EDATAINIT(site,
+                        ''.join("\nduplicate initializer for "
+                                + f"member '{field}'"
+                                for field in duplicates)
+                        + ''.join(f"\n'{field}' is not a member of "
+                                  + f"{etype}"
+                                  for field in bad_fields)
+                        + ("\nmissing initializer(s) for field(s) "
+                           + ", ".join(f"'{field}'"
+                                       for field in remaining)
+                           if remaining else '')
+                        )
+
+def loss_on_truncation(v, bits, signed):
+    if not math.isfinite(v) or int(v) != v:
```

Comment:
`math.isfinite` returns `False` for NaNs as well. Indeed, it's a safety measure in case float constants can be +/-Infinite or NaNs. From my experimentation, I actually haven't managed to create such a constant expression from DML -- but having this additional check costs nothing.